### PR TITLE
Enable the machine outliner for SyncVM

### PR DESCRIPTION
**Pull request by vladimirradosavljevic**

### CODE REVIEW

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.898 ║
║ Best                               45.104 ║
║ Worst                               0.000 ║
║ Total                               3.713 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                 -inf ║
║ Best                                0.000 ║
║ Worst                             -49.504 ║
║ Total                              -2.925 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                8.304 ║
║ Best                               31.070 ║
║ Worst                               0.000 ║
║ Total                              14.194 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                               -7.809 ║
║ Best                                0.000 ║
║ Worst                             -19.046 ║
║ Total                              -7.132 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.569 ║
║ Best                               37.020 ║
║ Worst                               0.000 ║
║ Total                               2.049 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                 -inf ║
║ Best                                0.000 ║
║ Worst                             -28.777 ║
║ Total                              -2.108 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                4.694 ║
║ Best                               31.070 ║
║ Worst                               0.000 ║
║ Total                               9.181 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                               -6.407 ║
║ Best                                0.000 ║
║ Worst                             -17.312 ║
║ Total                              -6.160 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.682 ║
║ Best                               37.020 ║
║ Worst                               0.000 ║
║ Total                               2.369 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                 -inf ║
║ Best                                0.000 ║
║ Worst                             -34.286 ║
║ Total                              -2.110 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                5.588 ║
║ Best                               31.070 ║
║ Worst                               0.000 ║
║ Total                              10.406 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                               -6.781 ║
║ Best                                0.000 ║
║ Worst                             -17.573 ║
║ Total                              -6.305 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.682 ║
║ Best                               37.020 ║
║ Worst                               0.000 ║
║ Total                               2.369 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                 -inf ║
║ Best                                0.000 ║
║ Worst                             -34.286 ║
║ Total                              -2.110 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                5.588 ║
║ Best                               31.070 ║
║ Worst                               0.000 ║
║ Total                              10.406 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                               -6.781 ║
║ Best                                0.000 ║
║ Worst                             -17.573 ║
║ Total                              -6.305 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.682 ║
║ Best                               37.020 ║
║ Worst                               0.000 ║
║ Total                               2.369 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                 -inf ║
║ Best                                0.000 ║
║ Worst                             -34.286 ║
║ Total                              -2.110 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                5.588 ║
║ Best                               31.070 ║
║ Worst                               0.000 ║
║ Total                              10.406 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                               -6.781 ║
║ Best                                0.000 ║
║ Worst                             -17.573 ║
║ Total                              -6.305 ║
╚═══════════════════════════════════════════╝

```

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (lialan) [inline llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp]: nit: define some macros for magic numbers to make it easier to read.

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (lialan) [inline llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp]: nit: again, can improve readability by avoid using magic numbers. same for the `CallOverhead` and `FrameOverhead` below.

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (lialan) [inline llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp]: nit: can use previous instruction's DebugLoc. Same for the one in `insertOutlinedCall`

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (lialan) [inline llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp]: should this be `succ_empty`?

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (lialan) [inline llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp:780]: I think you mean `hasImplicitUseOfPhyReg` here

**Review by lialan — CHANGES_REQUESTED** [review]
Good quality of code, just some places we need to revisit.

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (akiramenai) [reply llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp]: It shouldn't be.
But style: early return is preferred.

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (vladimirradosavljevic) [reply llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp]: Done.

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (vladimirradosavljevic) [reply llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp]: Done.

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (vladimirradosavljevic) [reply llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp]: Done.

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (vladimirradosavljevic) [reply llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp]: @lialan In case we don't have predecessors and MBB has no Flags instructions, we know that it is not propagating Flags across MBBs. Check for Flags instructions was missing, so I have added it, and updated test to reflect this possible issue.
@akiramenai Added early return.

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (vladimirradosavljevic) [reply llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp:780]: I'm looking for the Defs, not the Uses.
Please look at the following example:
```
...
Inst1 -> def Flags
....
Inst2 -> def Flags
....
jc dest, cc
```

We don't want to remove Inst1 from the list of candidates as we have Inst2 that defines flags and jc that uses it. Also, we want to remove Inst2.

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.684 ║
║ Best                               37.020 ║
║ Worst                               0.000 ║
║ Total                               2.371 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                 -inf ║
║ Best                                0.000 ║
║ Worst                             -34.286 ║
║ Total                              -2.112 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                5.588 ║
║ Best                               31.070 ║
║ Worst                               0.000 ║
║ Total                              10.406 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                               -6.781 ║
║ Best                                0.000 ║
║ Worst                             -17.573 ║
║ Total                              -6.305 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.684 ║
║ Best                               37.020 ║
║ Worst                               0.000 ║
║ Total                               2.371 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                 -inf ║
║ Best                                0.000 ║
║ Worst                             -34.286 ║
║ Total                              -2.112 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                5.588 ║
║ Best                               31.070 ║
║ Worst                               0.000 ║
║ Total                              10.406 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                               -6.781 ║
║ Best                                0.000 ║
║ Worst                             -17.573 ║
║ Total                              -6.305 ║
╚═══════════════════════════════════════════╝

```

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (lialan) [reply llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp:780]: Thanks. Maybe add this example to the comments above to show the intention.

Comment (vladimirradosavljevic) [reply llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp:780]: Sure, will do that.

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.684 ║
║ Best                               37.020 ║
║ Worst                               0.000 ║
║ Total                               2.371 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                 -inf ║
║ Best                                0.000 ║
║ Worst                             -34.286 ║
║ Total                              -2.112 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                5.588 ║
║ Best                               31.070 ║
║ Worst                               0.000 ║
║ Total                              10.406 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                               -6.781 ║
║ Best                                0.000 ║
║ Worst                             -17.573 ║
║ Total                              -6.305 ║
╚═══════════════════════════════════════════╝

```

File: .github/workflows/compiler-integration-tests.yml
Comment (lialan) [inline .github/workflows/compiler-integration-tests.yml:114]: yeah compile-twice has been broken for some tests for some while now.

And I think the upstream is not caring too much about it as well? @FlashSheridan

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (lialan) [inline llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp]: do remember to update this when checking in. The number hurts. :-)

File: llvm/test/CodeGen/SyncVM/machine-outliner.mir
Comment (lialan) [inline llvm/test/CodeGen/SyncVM/machine-outliner.mir:87]: c'est bon.

Just a comment to discuss: If flags are seen as ephemeral values, and it does not live (or we should not trust it) if we go beyond BB boundaries, then I am curious if machine verifier will complain line 147 to have a use-before-def of flags reg error, because verifier can work on the scope of a BB.

But what we have right now does no harm.

**Review by lialan — APPROVED** [review]
LGTM, just do remember to turn it off by default.

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (vladimirradosavljevic) [reply llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp]: Done.

**Comment by vladimirradosavljevic** [comment]
@lialan Thanks for a review!

File: llvm/test/CodeGen/SyncVM/machine-outliner.mir
Comment (vladimirradosavljevic) [reply llvm/test/CodeGen/SyncVM/machine-outliner.mir:87]: This is a good point, but for some reason, verifier is not complaining at all. I assume when we start implementing such optimizations, we will need to add Flags register in MBB liveins.

File: .github/workflows/compiler-integration-tests.yml
Comment (FlashSheridan) [reply .github/workflows/compiler-integration-tests.yml:114]: I’m afraid so:

> Filed https://github.com/llvm/llvm-project/issues/61462, no response yet, so not clear how serious the maintainers are about --compile-twice and/or asan.

https://linear.app/matterlabs/issue/CPR-1045#comment-03c2f235

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (akiramenai) [inline llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp:732]: Can we create a ticket which is blocked by the VM part (and put the number here)? If I remember right @sayon had raised it.

**Review by akiramenai — APPROVED** [review]
LGTM assuming flag handling logic will go off after switching to dynamic jumps.

File: llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp
Comment (vladimirradosavljevic) [reply llvm/lib/Target/SyncVM/SyncVMInstrInfo.cpp:732]: When we switch to `jump` implementation, we can remove this as we won't do `near_call` `ret` where this problem occurs, so comment wise, I suggest to leave it as is. What do you think?

@sayon Do we have a VM ticket for this issue?

**Comment by vladimirradosavljevic** [comment]
@akiramenai Yes, we will remove flag handling logic when we switch to jumps. Thanks for a review!

**Pull request merged**
